### ISSUE TITLE
NullPointerException in PartitionedLookupSourceFactory when using grouped execution and spill

### ISSUE THREAD

**Issue by findepi** [prestodb/presto#10258]
```
2018-03-26 14:28:36 SEVERE Error processing Split 20180326_084335_00012_gv34z.1.0-31  (start = 5.97478603262188E8, wall = 107 ms, cpu = 0 ms, wait = 116 ms, calls = 2)
java.lang.NullPointerException
	at com.facebook.presto.operator.PartitionedLookupSourceFactory$SpillAwareLookupSourceProvider.lambda$withLease$0(PartitionedLookupSourceFactory.java:430)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
	at com.facebook.presto.operator.PartitionedLookupSourceFactory$SpillAwareLookupSourceProvider.withLease(PartitionedLookupSourceFactory.java:430)
	at com.facebook.presto.operator.LookupJoinOperator.tryFetchLookupSourceProvider(LookupJoinOperator.java:236)
	at com.facebook.presto.operator.LookupJoinOperator.addInput(LookupJoinOperator.java:204)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:385)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:278)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:645)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:272)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:975)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

This is similar to (but distinct from) https://github.com/prestodb/presto/issues/10238

Repro steps:

1. Run `com.facebook.presto.hive.TestHiveIntegrationSmokeTest#testGroupedJoin`
2. Applying first the following to enable and enforce spilling:
```diff
diff --git presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java
index 0acceeca42..b8060afb6f 100644
--- presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java
+++ presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java
@@ -47,6 +47,7 @@ import java.util.function.Function;

 import static com.facebook.presto.SystemSessionProperties.COLOCATED_JOIN;
 import static com.facebook.presto.SystemSessionProperties.CONCURRENT_LIFESPANS_PER_NODE;
+import static com.facebook.presto.SystemSessionProperties.SPILL_ENABLED;
 import static com.facebook.presto.hive.HiveColumnHandle.BUCKET_COLUMN_NAME;
 import static com.facebook.presto.hive.HiveColumnHandle.PATH_COLUMN_NAME;
 import static com.facebook.presto.hive.HiveQueryRunner.HIVE_CATALOG;
@@ -2009,16 +2010,19 @@ public class TestHiveIntegrationSmokeTest
             // NOT grouped execution; default
             Session notColocated = Session.builder(getSession())
                     .setSystemProperty(COLOCATED_JOIN, "false")
+                    .setSystemProperty(SPILL_ENABLED, "true")
                     .build();
             // Co-located JOIN with all groups at once
             Session colocatedAllGroupsAtOnce = Session.builder(getSession())
                     .setSystemProperty(COLOCATED_JOIN, "true")
                     .setSystemProperty(CONCURRENT_LIFESPANS_PER_NODE, "-1")
+                    .setSystemProperty(SPILL_ENABLED, "true")
                     .build();
             // Co-located JOIN, 1 group per worker at a time
             Session colocatedOneGroupAtATime = Session.builder(getSession())
                     .setSystemProperty(COLOCATED_JOIN, "true")
                     .setSystemProperty(CONCURRENT_LIFESPANS_PER_NODE, "1")
+                    .setSystemProperty(SPILL_ENABLED, "true")
                     .build();

             assertQuery(notColocated, joinThreeBucketedTable, expectedQuery);
diff --git presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java
index 971f3d39ef..1272906945 100644
--- presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java
+++ presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java
@@ -86,14 +86,14 @@ public class FeaturesConfig
     private HistogramGroupImplementation histogramGroupImplementation = HistogramGroupImplementation.NEW;
     private boolean spillEnabled;
     private DataSize aggregationOperatorUnspillMemoryLimit = new DataSize(4, DataSize.Unit.MEGABYTE);
-    private List<Path> spillerSpillPaths = ImmutableList.of();
+    private List<Path> spillerSpillPaths = ImmutableList.of(Paths.get("/tmp/spill"));
     private int spillerThreads = 4;
     private double spillMaxUsedSpaceThreshold = 0.9;
     private boolean iterativeOptimizerEnabled = true;
     private boolean enableNewStatsCalculator;
     private boolean pushAggregationThroughJoin = true;
-    private double memoryRevokingTarget = 0.5;
-    private double memoryRevokingThreshold = 0.9;
+    private double memoryRevokingTarget = 0.0000001 / 2;
+    private double memoryRevokingThreshold = 0.0000001;
     private boolean parseDecimalLiteralsAsDouble;

     private Duration iterativeOptimizerTimeout = new Duration(3, MINUTES); // by default let optimizer wait a long time in case it retrieves some data from ConnectorMetadata
```

cc @haozhun

**Comment by findepi** [comment]
Spill has knowledge whether all join operators finished processing input. This is when unspilling begins. Before we start unspilling, we remove from memory any non-spilled partitions we may still have (https://github.com/prestodb/presto/blob/b4912717856d5c2750f16d0e521c37e8974fe406/presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java#L311).
To ensure that we indeed know when all join operators finish, we enforce that join operators are _not_ source distributed: https://github.com/prestodb/presto/blob/b4912717856d5c2750f16d0e521c37e8974fe406/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java#L448

Now, when probe side is grouped (bucket-by-bucket) and build side is not (i.e. when `SharedInternalLookupSourceFactoryManager` is used), the above assumption no longer holds -- join operators will be instantiated in groups (as many groups as there are buckets). Finishing of the first last join operator from the first group will trigger destruction of the PLSF (provided spilling occurred), rendering PLSF unusable for the next group.

**Comment by findepi** [comment]
I see following options:
1. disable spill in "probe bucketed, build not bucketed" case
2. when spill enabled, obliterate bucketing in "probe bucketed, build not bucketed"
   * in fact, this particular join doesn't benefit from bucketing, but i understand retaining bucketing might be useful for a join higher up the plan tree
3. when spill enabled, transforming a plan so that  "probe bucketed, build not bucketed" is the last join (and then obliterate bucketing there)
   * or the same, but without reordering -- obliterate bucketing in "probe bucketed, build not bucketed" if plan higher up doesn't benefit from this
4. any combination of the above, cost-based, for example if there are only 2 buckets, then bucketing doesn't increase our query capabilities significantly
   * however, 2 buckets is not a realistic case, in general, i have sense of feeling that bucketing is much stronger in terms of increasing query capabilities than the spill (when bucketing is applicable, if course)

From the above, my favorite picks are (combined):
* (3b) obliterating bucketing  in "probe bucketed, build not bucketed" case, when plan higher up doesn't benefit 
* (1) disable spill in the remaining "probe bucketed, build not bucketed" cases

@dain  @haozhun @sopel39 thoughts?

**Comment by sopel39** [comment]
Maybe we should initially prevent queries from failing via 1) and then think about more complete solution to the problem?
Seems that in `probe bucketed, build not bucketed` case we could disable spill locally and maintain bucketed output of lookup join operators. Would that be possible?

**Comment by sopel39** [comment]
> PLSF (provided spilling occurred), rendering PLSF unusable for the next group.

Can't we keep the partitions if we know probe side is bucketed?

**Comment by findepi** [comment]
> Can't we keep the partitions (for the next probe bucket to execute) if we know probe side is bucketed?

Spilling = not enough memory. In order to unspill the spilled partitions (and matching probe rows), we free memory by forgetting the partitions that has not been spilled (and are no longer needed).
If the "partitions that has not been spilled" are to be needed later (for next bucket of the probe side), we need to keep them around:
- spill them and then unspill?
- or keep in memory during unspilling? (This can work, but we need to free ~1 more partition for the unspilling -- this should happen automatically, via memory revocation)

So yes, seems doable.
